### PR TITLE
User-defined conversions for python bindings inside module source directory

### PR DESCRIPTION
**Pull request by ludv1x**
This PR adds more functionality to python bindings generator.
Now it is possible to define custom conversions between user defined C++ types and their Python analogous inside module source directory.
It helps yo avoid populating of python2/src/cv2.cpp and makes possible such conversions for contrib modules.
So, this PR is reqired for python bindings of [dnn module](https://github.com/ludv1x/opencv_contrib/tree/dnn/modules/dnn) from opencv_contrib.
@Nerei 

The second commit adds constructor and explicit destructors for "simple" classes inside wrapping code.
It improves safety and make able to wrap more complex classes (see comments for the diff).

### CODE REVIEW

File: modules/python/src2/gen2.py [hunk 1]
```diff
@@ -30,7 +30,7 @@
 gen_template_simple_call_constructor_prelude = Template("""self = PyObject_NEW(pyopencv_${name}_t, &pyopencv_${name}_Type);
         if(self) """)
 
-gen_template_simple_call_constructor = Template("""self->v = ${cname}${args}""")
```
Comment (ludv1x) [inline modules/python/src2/gen2.py]: Before such behavior made
`operator=()(<invalid object from python's malloc()>, Class(args))`
Now it directly constucts object on allocated data using placement new, no extra assigment is perfomed.

File: modules/python/src2/gen2.py [hunk 2]
```diff
@@ -74,6 +74,7 @@
 
 static void pyopencv_${name}_dealloc(PyObject* self)
 {
+    ((pyopencv_${name}_t*)self)->v.${cname}::~${sname}();
```
Comment (ludv1x) [inline modules/python/src2/gen2.py]: Before there are no explicit destructor.
It is ok and fast for trivial structs, but unacceptable for more complex classes.
As for performance, compiler's optimizations should elimnate possible overhead on trival classes.

File: modules/python/src2/gen2.py [hunk 3]
```diff
@@ -74,13 +74,14 @@
...
     PyObject_Del(self);
 }
 
 template<> PyObject* pyopencv_from(const ${cname}& r)
 {
     pyopencv_${name}_t *m = PyObject_NEW(pyopencv_${name}_t, &pyopencv_${name}_Type);
-    m->v = r;
```
Comment (ludv1x) [inline modules/python/src2/gen2.py]: Copy constructor `Class(r)` instead of
`operator = (<invalid state obj>, r)

**Comment by vpisarev** [comment]
@mshabunin, I think, this is quite important patch; could you please review it?

**Comment by alalek** [comment]
Looks good to me! :+1:

**Comment by mshabunin** [comment]
@ludv1x , we already have `<module>/misc` location for same purpose (currently only _java_ bindings use it https://github.com/opencv/opencv/tree/master/modules/core/misc/java). Is it possible to move manual-written headers for python to same folder (`<module>/misc/python`)?

**Comment by ludv1x** [comment]
@mshabunin Yes, of course. Good idea.

**Comment by ludv1x** [comment]
@mshabunin @vpisarev So, can we merge it now?

**Comment by alalek** [comment]
@ludv1x Please squash all commits into one.

**Comment by ludv1x** [comment]
@alalek Ok.

**Comment by mshabunin** [comment]
:+1:

**Pull request merged**